server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/reactivestreams/SimpleSubscriber.scala

```diff
@@ -6,17 +6,19 @@ import org.reactivestreams.{Publisher, Subscription}
 import sttp.capabilities.StreamMaxLengthExceededException
 
 import java.util.concurrent.LinkedBlockingQueue
-import scala.concurrent.{Future, Promise}
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, Promise}
 
-private[netty] class SimpleSubscriber(contentLength: Option[Int]) extends PromisingSubscriber[Array[Byte], HttpContent] {
+private[netty] class SimpleSubscriber(contentLength: Option[Long]) extends PromisingSubscriber[Array[Byte], HttpContent] {
+  // These don't need to be volatile as Reactive Streams guarantees that onSubscribe/onNext/onError/onComplete are
+  // called serially (https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#1-publisher-code - rule 3)
```

Comment:
**question:** Isn't the serial guarantee a different thing than memory barrier concerns? We have a guarantee that there will be no parallel access to these values, but they can still happen on different threads, thus a change performed on thread may not be always immediately visible to another thread?

Reply:
AFAIR, it includes memory visibility safety (the "happens-before relationship"):

The definition of "serially" says:

> In the context of a [Signal](https://github.com/reactive-streams/reactive-streams-jvm#term_signal), non-overlapping. In the context of the JVM, calls to methods on an object are serial if and only if there is a happens-before relationship between those calls (implying also that the calls do not overlap). When the calls are performed asynchronously, coordination to establish the happens-before relationship is to be implemented using techniques such as, but not limited to, atomics, monitors, or locks.

Reply:
I understood this as a guarantee that calls to onNext, onComplete, etc. never overlap, so we are safe to perform check internal state and perform updates without worrying that these values will be changed in a parallel onNext call. I think you may be right regarding these guarantees extending to visibility, at least internally. For example, checking some other open source subscribers: https://github.com/playframework/netty-reactive-streams/blob/89aff754154a118ea134d5ca64cce3dc44c78fc6/netty-reactive-streams/src/main/java/org/playframework/netty/HandlerSubscriber.java#L84
Most mutable members are normal variables. The `Subscription` and `ChannelHandlerContext` are volatile, possibly because they are provided from the outside?

Reply:
> I understood this as a guarantee that calls to onNext, onComplete, etc. never overlap, so we are safe to perform check internal state and perform updates without worrying that these values will be changed in a parallel onNext call.

Without proper happens-before relationship this "non-overlapping" alone doesn't really give you any safety at all because even though subsequent `onNext` calls happen one after another in real time, they can still see some not-fully-thread-published broken state of a previous invocation, requiring the same synchronization as if they were concurrent.

> Most mutable members are normal variables. The Subscription and ChannelHandlerContext are volatile, possibly because they are provided from the outside?

I think `subscription` must be `volatile` because it is accessed from `ChannelDuplexHandler` handler methods. There is no happens-before relationship between `onSubscribe` (where `subscription` is set) and these methods.

You have a similar situation with `ctx`, which is set by a Netty handler method but accessed in `Subscriber` methods.

Reply:
Also, the `HandlerSubscriber` is a Netty channel handler, so its `onNext` and other reactive methods are supposed to be called only from Netty handler context, which guarantees the same thread, thus visibility. In case of our `SimpleSubscriber` it's similar - it is the Netty's underlying publisher that calls these methods, and `totalLength` and `buffers` are accessed only there, so I guess it's OK to remove `volatile`.

Reply:
Yeah, I guess this is the way Netty implements reactivestreams spec's requirements.